plat/drivers/balloon/balloon_drv.c

```diff
@@ -0,0 +1,372 @@
+		vb->transport->num_pages -= 1;
+	}
+}
+
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+ *
+ * Copyright (c) 2011, Bryan Venteicher <bryanv@FreeBSD.org>
+ * All rights reserved.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* The above copyright notice applies only to the below function,
+ * vtballoon_send_page_frames, which is based on FreeBSD's function
+ * of the same name.
+ */
+
+static void vtballoon_send_page_frames(struct virtio_balloon_device *vb,
+			struct virtqueue *vq, int npages)
+{
+	struct uk_sglist sg;
+	struct uk_sglist_seg segs[1];
+	int c;
+	void *vq_cookie;
+	__u32 len = 0;
+
+	uk_sglist_init(&sg, 1, segs);
+
+	uk_sglist_append(&sg, vb->transport->pages, npages * sizeof(uint32_t));
+
+	virtqueue_buffer_enqueue(vq, &vq_cookie, &sg, 1, 0);
+
+	virtqueue_host_notify(vq);
+
+	/* wait on KVM to respond. Need a safer method for this */
+	while ((c = virtqueue_buffer_dequeue(vq, &vq_cookie, &len)) < 0)
+		;
+
+}
+
+/**
+ * This is equivalent to leaking from the balloon and
+ * increasing memory reservation for guest.
+ */
+int deflate_balloon(uintptr_t *pages_to_guest, uint32_t num)
+{
+	struct virtio_balloon_device *vb = global_vb;
+	int num_pages_taken;
+	uint32_t i;
+
+	/* check if device is ready */
+	if (!global_vb)
+		return -ENXIO;
+
+	uk_mutex_lock(&vb->lock);
+
+	clear_transport(vb);
+
+	if (vb->balloon->num_pages < num)
+		num = vb->balloon->num_pages;
+
+	for (i = 0; i < num; i++) {
+		vb->transport->pages[i] = pages_to_guest[i];
+		vb->balloon->num_pages -= 1;
+		vb->transport->num_pages += 1;
+	}
+
+	num_pages_taken = vb->transport->num_pages;
+
+	if (vb->transport->num_pages != 0) {
+		vtballoon_send_page_frames(vb, vb->deflate_vq,
+			vb->transport->num_pages);
+	}
+
+	uk_mutex_unlock(&vb->lock);
+
+	return num_pages_taken;
+}
+
+/**
+ * This is equivalent to filling the balloon and
+ * decreasing memory reservation for guest.
+ */
+int inflate_balloon(uintptr_t *pages_to_host, uint32_t num)
+{
+	struct virtio_balloon_device *vb = global_vb;
+	int num_pages_given;
+	uint32_t i;
+
+	/* check if device is ready */
+	if (!global_vb)
+		return -ENXIO;
+
+	uk_mutex_lock(&vb->lock);
+
+	clear_transport(vb);
+
+	for (i = 0; i < num; i++) {
+		vb->transport->pages[i] = pages_to_host[i] / __PAGE_SIZE;
```

Comment:
Are you storing some sort of page index into `vb->transport->pages[i]`? I am wondering about this division. Can you add a comment?